topic_popover: Add support to follow a topic from the topic_popover.

prakhar1144 opened the pull request:
<!-- Describe your pull request here.-->
### First commit
* Rename `muted_topics_ui.js` to `user_topics_ui.js`
---
### Second commit:
- Adds the missing node tests for the case when the visibility policy is UNMUTED. 
---
### Third commit:
*   Fix the live-update bug in the 'SETTINGS / TOPICS' panel. -- [details](https://github.com/zulip/zulip/pull/26016#pullrequestreview-1481765759)
---
### Fourth commit
**Update SETTINGS / TOPICS panel to include followed topics.**
- [x] Add "Followed" as the first option in the topic status dropdowns.
- [x] Modify the intro text.
- [x] Replace the "Learn more." link with a question mark linking to /help/mute-a-topic on the "Topic settings" heading.
---
### Fifth commit
**Follow a topic from the left sidebar three-dot menu.**
- Always show three options at the top of the menu, regardless of whether the stream is muted or unmuted.
- There is no way to reset the topic to "default" from this menu -- that can only be done via the Topics settings panel.

**Pending Discussion:**
- Should we highlight the selected option and make it non-clickable? -- [CZO discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/Follow.20Topics/near/1593108)
- Follow Topic icon
---
### Sixth & Seventh commit
Remaining parts of #25640 
- [x] Followed Topic desktop notifications
- [x] Followed Topic audible notifications
---

Fixes: #25918 
Fixes: #25917 

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
<details>
<summary>
<b>SETTINGS/TOPIC panel</b>
</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/54594283-b023-4f26-85d6-0a11eed21967"></img>
</details>
<details>
<summary>
<b>Topic popover</b>
</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/711e9b98-5b94-4c2a-8b41-b3e0ead783c2"></img>
</details>
<hr>
<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

prakhar1144 commented:
@zulipbot add "buddy review"

abhijeetbodas2001 commented:
Do we want to add a "follow topic" option in the recipient bar?

abhijeetbodas2001 commented on web/tests/notifications.test.js:
```diff
@@ -119,7 +126,31 @@ test("message_is_notifiable", () => {
+        topic: "followed topic",
+    };
+    assert.equal(notifications.should_send_desktop_notification(message), true);
+    assert.equal(notifications.message_is_notifiable(message), true);
```
We should also verify here that for normal topics, `message_is_notifiable == false`

prakhar1144 replied:
I think this test already covers all the cases where `message_is_notifiable` should be `false`.

For normal topics (not muted and no visibility policy set), `message_is_notifiable` will be `true`, and should_send desktop/audible notifications will be determined by the respective settings. (This is on line 285)

abhijeetbodas2001 reviewed (commented):
I tested this out manually, here are some notes:
1. Open two tabs. In the first one, open `/#settings/topics`. From the other tab, mute/unmute/follow a topic. This change is not reflected in the settings pane in the other tab. I don't know if this was the case from earlier, or is a regression introduced recently. But in any case I think we should fix it.
2. If topic `foo` is UNMUTED (and shown as such in the settings pane), if you FOLLOW it from one tab, the value isn't updated in the dropdown in the settings pane in the other tab. This might be related to (1) as a live-update bug.
3. If a topic is MUTED (or UNMUTED or FOLLOWED), we still show the option the MUTE it in the topic popover. I think we should hide that option and only show the other options (and probably add the "stream default option")
4. The desktop/audible notifications seem to be working well :+1:

prakhar1144 commented:
> Do we want to add a "follow topic" option in the recipient bar?

Yes, This is planned as a part of #25915

prakhar1144 commented:
> 1. Open two tabs. In the first one, open /#settings/topics. From the other tab, mute/unmute/follow a topic. This change is not reflected in the settings pane in the other tab. I don't know if this was the case from earlier, or is a regression introduced recently. But in any case I think we should fix it.
> 2. If topic foo is UNMUTED (and shown as such in the settings pane), if you FOLLOW it from one tab, the value isn't updated in the dropdown in the settings pane in the other tab. This might be related to (1) as a live-update bug.

This is an interesting case. In #25127, we removed [this code block](https://github.com/zulip/zulip/pull/25127/files#diff-5f0b4dc6f8f414b16f672e561e195319e08527343866e3ed9b13e22dbf7027a5).
The [expected behaviour](https://github.com/zulip/zulip/pull/25127#issuecomment-1513760715) was:
```
2. Let's not have the topics get resorted when you change the status in this UI, as it makes it harder to
find what you were looking at. The order should just get updated when you navigate away or refresh.

3. When you set a topic to "Default for stream", it should stay in the UI until you refresh or navigate away.
This makes it much easier to review your changes, and undo if needed. It's the same pattern we follow in
the Users panel when an admin deactivates a user, for example.
```
**I have fixed the live-update bug in the first commit.** 

Note: Either we change the status from topic popover or from the panel dropdown, only the status will be updated (It won't get re sorted by date updated or removed on changing the status to "default for stream").

abhijeetbodas2001 commented:
> Note: Either we change the status from topic popover or from the panel dropdown, only the status will be updated (It won't get re sorted by date updated or removed on changing the status to "default for stream").

Great! I think this is a fine tradeoff for now, but please also bring this up in the CZO topic for discussion.

> ICYMI: I have already started a discussion regarding this on [CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/Follow.20Topics/near/1593108) :) (Earlier, I mentioned this in the PR description.)

Indeed I missed it :) Thanks for the link, I'l  bump the topic.

abhijeetbodas2001 commented on web/src/muted_topics_ui.js:
```diff
@@ -1,6 +1,10 @@
 import * as recent_topics_ui from "./recent_topics_ui";
+import * as settings_user_topics from "./settings_user_topics";
 import * as stream_list from "./stream_list";
 import * as sub_store from "./sub_store";
```
This file is still called `muted_topics_ui.js`. I think we should change that to `user_topics_ui.js` like we've done for the other related files.

prakhar1144 replied:
Makes sense. Addressed this in the 7th commit.

prakhar1144 commented:
Updates in the recent push:
* Fixed a bug related to desktop/audible notifications in followed topic.
  * With the following settings configuration, if we send a message having wildcard mention in a followed topic, desktop + audible notifications should be generated.
  * Fixed this bug: [Related code](https://github.com/zulip/zulip/pull/26016/files#diff-b87d2b5cdd54710b9a58fc1db373a5ac2241e77da9802cf08f46168adb31d3caR424) + [Related test](https://github.com/zulip/zulip/pull/26016/files#diff-1b67a80ff323893d633d85363c68445ebde063cadd46b3d173d0cd58435825eaR272).

* Addressed [#26016 (comment)](https://github.com/zulip/zulip/pull/26016#issuecomment-1596107343)
  * > Great! I think this is a fine tradeoff for now, but please also bring this up in the CZO topic for discussion.
  * Started a [CZO discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/Follow.20Topics/near/1598161)

* Renamed `muted_topics_ui.js` to `user_topics_ui.js` -- 7th commit

Awaiting feedback:
* Follow Topics icon -- [CZO discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/Follow.20Topics/near/1593012)
* Settings / Topics panel live update tradeoff -- [CZO discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/Follow.20Topics/near/1598161)
* Topic popover selected option highlight or remove -- [CZO discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/Follow.20Topics/near/1593108)

prakhar1144 commented on web/templates/user_topic_ui_row.hbs:
```diff
@@ -5,7 +5,9 @@
         <select class="settings_user_topic_visibility_policy prop-element list_select bootstrap-focus-style" data-setting-widget-type="number">
             {{#each ../user_topic_visibility_policy_values}}
-                <option value='{{this.code}}' {{#if (eq this.code ../visibility_policy)}}selected{{/if}}>{{this.description}}</option>
+                {{#unless (and (eq this.description "Followed") (not ../../development))}}
+                    <option value='{{this.code}}' {{#if (eq this.code ../visibility_policy)}}selected{{/if}}>{{this.description}}</option>
+                {{/unless}}
```
I think we will still need the calculation. We will still iterate over `user_topic_visibility_policy_values` and when `description==Followed`; we will check for `show_followed_in_dropdown`.

Anyways, this is a temporary change; we will remove this calculation when we make the feature available to all the users.

abhijeetbodas2001 replied:
Cool, makes sense!

prakhar1144 commented on web/src/notifications.js:
```diff
@@ -403,6 +421,15 @@ export function should_send_desktop_notification(message) {
+    // Followed topic wildcard mentions
+    if (
+        message.mentioned &&
+        user_topics.is_topic_followed(message.stream_id, message.topic) &&
+        user_settings.enable_followed_topic_wildcard_mentions_notify
```
> Would it be a good idea to merge this logic into the `stream_data.receives_notifications` function which is called above?

No, because it deals with prioritizing the stream-specific setting (if exists) over the global notification setting for a stream.
> if there is any scope to centralize the user-topic logic similar to the stream notifications logic.

I think it would make more sense to centralize it in a similar way, when we will have global and topic-specific setting kind of thing.

`stream_data.receives_notifications` has been used at 3 more places (not counting tests); but the intention at all the places is to find:
* is stream_specific and default setting dissimilar
* fetching the stream-specific setting over global

abhijeetbodas2001 replied:
Great, makes sense to have the topic code separate then :+1:

prakhar1144 commented on web/src/notifications.js:
```diff
@@ -403,6 +421,15 @@ export function should_send_desktop_notification(message) {
         return true;
     }
 
+    // Followed topic wildcard mentions
+    if (
+        message.mentioned &&
+        user_topics.is_topic_followed(message.stream_id, message.topic) &&
+        user_settings.enable_followed_topic_wildcard_mentions_notify
+    ) {
+        return true;
+    }
+
```
Comment added.

prakhar1144 commented on web/tests/notifications.test.js:
```diff
@@ -128,8 +129,8 @@ test("message_is_notifiable", () => {
     // Case 4: If the message has been sent to a followed topic,
-    // DO visually notify the user if 'enable_followed_topic_desktop_notifications'
-    // is enabled.
+    // DO visually and aurally notify the user if 'enable_followed_topic_desktop_notifications'
```
**aurally** is also correct; changed it to **audibly** for simplicity.

prakhar1144 commented on tools/test-js-with-node:
```diff
@@ -230,6 +229,7 @@ EXEMPT_FILES = make_set(
         "web/src/user_status.js",
         "web/src/user_status_ui.js",
         "web/src/user_topics.js",
+        "web/src/user_topics_ui.js",
```
Done.

prakhar1144 commented on web/tests/user_topics.test.js:
```diff
@@ -160,6 +160,49 @@ test("get_mutes", () => {
     ]);
 });
 
+test("get_unmutes", () => {
+    assert.deepEqual(
+        user_topics.get_user_topics_for_visibility_policy(
+            user_topics.all_visibility_policies.UNMUTED,
+        ),
+        [],
```
Done.

prakhar1144 commented:
As [discussed in CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/Follow.20Topics.20UI/near/1598451):


[topic-popover.webm](https://github.com/zulip/zulip/assets/56781761/84c7aa76-48a6-4b69-9855-f306be9b0b0c)

<details><summary>Unmuted</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/b2874c63-0427-49c0-b5fa-ae61f2bd37cc">
</img>
</details> 

<details><summary>Default</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/30ac19a9-22cc-4aa0-b35c-eb0b8fc3a636">
</img>
</details> 

<details><summary>Tooltip</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/f7c47ded-8ffe-42a0-8406-8b699f683e26">
</img>
</details> 

<details><summary>Tooltip - Light Theme</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/e5a86e0c-39f4-46e3-a262-d7b5f5889431">
</img>
</details> 

<details><summary>Unmuted - Light Theme</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/d8f067fa-6780-458a-a7d5-8211e4da7a73">
</img>
</details>

prakhar1144 commented on web/src/popover_menus_data.js:
```diff
@@ -144,5 +144,6 @@ export function get_topic_popover_content_context({stream_id, topic_name, url})
         color: sub.color,
         has_starred_messages,
         url,
+        development: page_params.development_environment,
```
Since, we have a complete different UI (tabs) now; please have a look at the 5th commit.

abhijeetbodas2001 commented:
I haven't looked at the CSS commit here since I don't have much experience on that front :sweat_smile: 
Maybe @amanagr could you take a look at that commit?
The rest of the commits look good to me (I looked at the code, as well as manually tested the PR).

One UX question -- should we be closing the topic popover after changing the visibility policy from the tabs? It seems a bit unnatural to me. For example, I am using a Gnome 43.4 desktop, which has a notifications popup (image below). When I enabled the DND mode (which has a similar tab like UI which we're trying to implement here), the notifications popup does not close.
![image](https://github.com/zulip/zulip/assets/55339528/bf9b54b9-499c-4ae7-9a88-05c1830866fd)

terpimost commented:
I think it is ok to close popover on change of muting. Later we should change the background of the tooltip in general... I think somewhere there was a work on that track involving personal menu change.

Icons are provided here https://chat.zulip.org/#narrow/stream/101-design/topic/Follow.20Topics.20UI/near/1604143

prakhar1144 commented on web/tests/user_topics.test.js:
```diff
@@ -139,12 +139,56 @@ test("get_mutes", () => {
     ]);
 });
 
+test("get_unmutes", () => {
+    assert.deepEqual(
+        user_topics.get_user_topics_for_visibility_policy(
```
Done

prakhar1144 commented on web/src/user_topics_ui.js:
```diff
@@ -21,6 +25,26 @@ export function handle_topic_updates(user_topic_event) {
+            $status.val(visibility_policy);
+        } else {
+            settings_user_topics.populate_list();
+        }
```
Comment added

```javascript
// If the row doesn't exist, the user must have set the visibility policy
// via another tab. We call 'populate_list' to re-create the panel, hence
// including the new row.
```

prakhar1144 commented on web/src/click_handlers.js:
```diff
@@ -415,7 +412,7 @@ export function initialize() {
     // Mute topic in a muted stream
     $("body").on("click", ".message_header .stream_muted.on_hover_topic_mute", (e) => {
         e.stopPropagation();
-        muted_topics_ui.mute_or_unmute_topic(
+        user_topics_ui.mute_or_unmute_topic(
             $(e.target),
```
`mute_or_unmute_topic` is related to `message_header` + `recent_topics_table`. Should we rename this here since we will replace the current UI with the proposed tabs UI? We're finally going to remove this function altogether (once the feature is live)

proposed UI: https://chat.zulip.org/#narrow/stream/101-design/topic/Follow.20Topics.20UI/near/1599171

---

After discussion in CZO (DM) with @abhijeetbodas2001 ; added a comment in the file:
```javascript
// This function is only called from the message header & recent_topics table
// and will be removed once we migrate to the new UI for selecting visibility policy
// for a topic.
export function mute_or_unmute_topic($elt, visibility_policy) {
...
}

prakhar1144 commented:
As [discussed in CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/Follow.20Topics.20UI/near/1598451):


[updated-ui.webm](https://github.com/zulip/zulip/assets/56781761/d2adede8-d61f-4ae2-9007-e8bbb6d19dee)

<details><summary>Light Theme</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/e6e17080-9b3f-4178-bbe6-b33ca968c235">
</img>
</details> 

Updates in the recent push:
* Used new icons
* Don't use black color for icons in the light theme
* Addressed review comments

terpimost commented:
I suggest to reorder items, so muted will be on the left and follow will be on the right

prakhar1144 commented:
@terpimost, made the changes.

<details><summary>Dark Theme</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/a9920212-9c46-4737-99c0-057800d6b4d3">
</img>
</details> 

<details><summary>Light Theme</summary>
<img src="https://github.com/zulip/zulip/assets/56781761/16db0ad3-410f-420c-b018-5b9623cddbb6">
</img>
</details>

terpimost commented:
great

prakhar1144 commented:
@zulipbot add "mentor review"

prakhar1144 commented:
@zulipbot remove "mentor review"

prakhar1144 commented:
@zulipbot add "maintainer review"

prakhar1144 commented:
The only thing left to review here is the CSS commit [(5th commit)](https://github.com/zulip/zulip/pull/26016/commits/2c028ae5fd503255ccd379e3994fef0a2423a5fd).

ref: [#26016 (Comment)](https://github.com/zulip/zulip/pull/26016#issuecomment-1625481760)

timabbott commented on web/templates/settings/user_topics_settings.hbs:
```diff
@@ -1,14 +1,27 @@
+            {{#tr}}
+                Zulip lets you configure notifications for each topic. You can also automatically follow or unmute topics
+                you start or participate in.
+            {{/tr}}
```
This copy probably isn't what we'll want long-term; linking to /help/ documentation seems valuable, and we're also not exactly configuring notifications directly; we're configuring notifications policy for the topic at a somewhat higher level...

timabbott commented on web/styles/dark_theme.css:
```diff
@@ -1543,6 +1543,26 @@
+                border: 1px solid hsl(0deg 0% 100% / 12%);
+            }
+        }
+    }
```
This should be discussed in #design and can be fixed in a follow-up, but playing with this, it feels a bit weird that none of these options are selected for a "default configuration" topic.

prakhar1144 replied:
[CZO Discussion](https://chat.zulip.org/#narrow/stream/101-design/topic/Follow.20Topics.20UI/near/1614625)

prakhar1144 replied:
#26486

timabbott commented on web/styles/zulip.css:
```diff
@@ -3189,3 +3189,43 @@ select.invite-as {
     }
 }
+
+.tabs-container {
```
I think this probably should go in `app_components.css`; we generally put fairly well-defined reusable components there.

timabbott replied:
I moved this while resolving the merge conflicts.

timabbott commented:
I think the settings "topics" UI is missing a loading...saved indicator interaction to show you that your changes are saved; right now you get no feedback at all that your change works. But that feels like something that'd be OK as an immediate follow-up.

This looks like great progress, marked to merge once CI passes after resolving the merge conflicts. Posted a few comments above on other things we'll want to tweak before making this visible outside the development environment. Huge thanks for all the work on this @prakhar1144 and @abhijeetbodas2001!
